accounts/store.go

```diff
@@ -0,0 +1,311 @@
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, dbPathPermission); err != nil {
+			return nil, err
+		}
+	}
+
+	// Open the database that we'll use to store the primary macaroon key,
+	// and all generated macaroons+caveats.
+	path := filepath.Join(dir, fileName)
+	db, err := bbolt.Open(path, dbFilePermission, &bbolt.Options{
+		Timeout: DefaultAccountDBTimeout,
+	})
+	if err == bbolt.ErrTimeout {
+		return nil, fmt.Errorf("error while trying to open %s: timed "+
+			"out after %v when trying to obtain exclusive lock",
+			path, DefaultAccountDBTimeout)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// If the store's bucket doesn't exist, create it.
+	err = db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(accountBucketName)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Return the DB wrapped in a BoltStore object.
+	return &BoltStore{db: db}, nil
+}
+
+// Close closes the underlying bolt DB.
+func (s *BoltStore) Close() error {
+	return s.db.Close()
+}
+
+// NewAccount creates a new OffChainBalanceAccount with the given balance and a
+// randomly chosen ID.
+func (s *BoltStore) NewAccount(balance lnwire.MilliSatoshi,
+	expirationDate time.Time) (*OffChainBalanceAccount, error) {
+
+	if balance == 0 {
+		return nil, fmt.Errorf("a new account cannot have balance of 0")
+	}
+
+	// First, create a new instance of an account. Currently, only the type
+	// OneTimeBalance is supported.
+	id, err := s.uniqueRandomAccountID()
+	if err != nil {
+		return nil, err
+	}
+	account := &OffChainBalanceAccount{
+		ID:             id,
+		Type:           OneTimeBalance,
+		InitialBalance: balance,
+		CurrentBalance: int64(balance),
+		ExpirationDate: expirationDate,
+		Invoices:       make(map[lntypes.Hash]struct{}),
+		Payments:       make(map[lntypes.Hash]*PaymentEntry),
+	}
+
+	// Try storing the account in the account database, so we can keep track
+	// of its balance.
+	if err := s.UpdateAccount(account); err != nil {
+		return nil, err
+	}
+	return account, nil
+}
+
+// UpdateAccount writes an account to the database, overwriting the existing one
+// if it exists.
+func (s *BoltStore) UpdateAccount(account *OffChainBalanceAccount) error {
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(accountBucketName)
+		if bucket == nil {
+			return ErrAccountBucketNotFound
+		}
+
+		account.LastUpdate = time.Now()
+		accountBinary, err := serializeAccount(account)
+		if err != nil {
+			return err
+		}
+
+		return bucket.Put(account.ID[:], accountBinary)
+	})
+}
+
+// uniqueRandomAccountID generates a new random ID and makes sure it does not
+// yet exist in the DB.
+func (s *BoltStore) uniqueRandomAccountID() (AccountID, error) {
+	var newID AccountID
+	if _, err := rand.Read(newID[:]); err != nil {
+		return newID, err
+	}
+
+	if _, err := s.Account(newID); err != ErrAccNotFound {
+		return s.uniqueRandomAccountID()
```

Comment:
Ah ok, we'll just retry....👍

Though I wonder if a loop is better to batter manage the stack size? 

On the otherhand, don't really imagine this needing to be executed this many times given the large range.

Reply:
Also feels like we could bundle this entire operation in a DB transaction as well (account create, retry, commit).

Reply:
Yeah, makes sense to limit the number of attempts and do everything in a single transaction. I wrote some of this code back in 2018/2019, good to get another pair of eyes to take a look at it :sweat_smile: